semantic-ast/app/Main.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Main (main) where
+
+import AST.GenerateSyntax
+import Control.Lens (Traversal', mapped, (%~))
+import Control.Monad
+import Data.Generics.Product.Typed (typed)
+import Data.Maybe
+import Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Foreign
+import GHC.Generics (Generic)
+import Language.Haskell.TH hiding (JavaScript)
+import Language.Haskell.TH.Lens
+import NeatInterpolation
+import qualified Options.Generic as Opt
+import Source.Language
+import System.Directory
+import System.Exit
+import qualified TreeSitter.Language
+import System.IO
+import System.Process
+import qualified TreeSitter.Go as Go (tree_sitter_go)
+import qualified TreeSitter.JSON as JSON (tree_sitter_json)
+import qualified TreeSitter.Java as Java (tree_sitter_java)
+import qualified TreeSitter.PHP as PHP (tree_sitter_php)
+import qualified TreeSitter.Python as Python (tree_sitter_python)
+import qualified TreeSitter.QL as CodeQL (tree_sitter_ql)
+import qualified TreeSitter.Ruby as Ruby (tree_sitter_ruby)
+import qualified TreeSitter.TSX as TSX (tree_sitter_tsx)
+import qualified TreeSitter.TypeScript as TypeScript (tree_sitter_typescript)
+
+data Config = Config {language :: Text, path :: FilePath}
+  deriving stock (Show, Generic)
+  deriving anyclass (Opt.ParseRecord)
+
+-- There are a few cases where the output emitted by TH's 'pprint' doesn't
+-- create entirely valid Haskell syntax, because sometimes we get
+-- a qualified name on the LHS of a typeclass declaration, which Haskell
+-- doesn't like at all. I haven't figured out quite why we get this qualified
+-- name, but for now the easiest thing to do is some nested updates with lens.
+adjust :: Dec -> Dec
+adjust = _InstanceD . typed . mapped %~ (values %~ truncate) . (functions %~ truncate)
+  where
+    -- Need to handle functions with no arguments, which are parsed as ValD entities,
+    -- as well as those with arguments, which are FunD.
+    values, functions :: Traversal' Dec Name
+    values = _ValD . typed . _VarP
+    functions = _FunD . typed
+
+    truncate :: Name -> Name
+    truncate = mkName . nameBase
```

Comment:
As always, there’s a dilemma between tedious/unreadable pattern matching and cryptic/quasi-readable lens operators. I went with the latter, since we already use `lens` in the project, and it comes with lenses/prisms for the types we need. (TH doesn’t define record selectors for its product types, so you can’t even do record updates.)